WordPress/gutenberg
# DataForm: implement first prototype using duplicate page action

Part of https://github.com/WordPress/gutenberg/issues/59745

## What?

First step in implementing the `DataForm` API. It's only used by the duplicate action from pages, which is only available in the plugin.

<img width="406" alt="Captura de ecrã 2024-07-01, às 18 35 34" src="https://github.com/WordPress/gutenberg/assets/583546/567503e9-358e-422c-82de-2b5b128e8599">

## Why?

We want to start exploring this API for several aspects of our flows. The first step is in replacing the "Duplicate item" action for pages because:

- That action is only available in the plugin, providing a nice place to hide this
- After merging, there's an organic path for expanding this API:
  - Rename actions for page and custom patterns. It's the same as this one.
  - Duplicate action for parts. It requires a new component (list for area).
  - Duplicate action for patterns. It requires adding two new components (boolean for synced, labels for categories).
  - etc.

| Page | Template part  | Pattern |
| ----- | --------------  | -------- |
| <img width="406" alt="Captura de ecrã 2024-07-01, às 18 35 34" src="https://github.com/WordPress/gutenberg/assets/583546/567503e9-358e-422c-82de-2b5b128e8599"> | <img width="510" alt="240701-duplicate-action-parts" src="https://github.com/WordPress/gutenberg/assets/583546/6907ff55-e2f8-4675-b52a-d28a3b363e67"> | <img width="510" alt="240701-duplicate-action-patterns" src="https://github.com/WordPress/gutenberg/assets/583546/e53d956a-5977-48ba-8140-b5d34788cdf2"> |

## How?

- Creates a `DataForm` component that temporarily lives within the `@wordpress/dataviews` until we find a better place.
- Refactors the duplicate action for pages to use the new component.

## Testing Instructions

- Go to "Site Editor > Pages". In any page, execute the "Duplicate action" and verify it works.
- Edit a page using any editor (Post or Site). Execute the "Duplicate action" and verify it works.

## Follow-ups

- Extract field definition to a common place (registry?) that can be accessed by the actions & the pages components. https://github.com/WordPress/gutenberg/pull/63849
- Expand DataForm to other actions (see example above).
- Storybook story. https://github.com/WordPress/gutenberg/pull/63840
